lib/Dialect/StableHLO/Utils/ShardyUtils.cpp

```diff
@@ -95,66 +73,139 @@ inline MeshMap createMeshMapFromMeshAttr(mlir::sdy::MeshAttr meshAttr) {
-
-      if (result.wasInterrupted()) {
-        return true;
+    // Check if sdy.sharding exists for any of the operations.
+    mlir::WalkResult result = funcOp.getBody().walk([&](mlir::Operation *op) {
+      if (auto currentArgAttrDict =
+              mlir::DictionaryAttr::get(module.getContext(), op->getAttrs())) {
+        if (currentArgAttrDict.contains(mlir::sdy::TensorShardingAttr::name)) {
+          return WalkResult::interrupt();
+        }
       }
+
+      return WalkResult::advance();
+    });
+
+    if (result.wasInterrupted()) {
+      return true;
     }
   }
 
   return false;
 }
 
 // Check if the module has any gspmd annotations.
-inline bool gspmdAnnotationsExist(mlir::ModuleOp &module) {
+bool gspmdAnnotationsExist(mlir::ModuleOp &module) {
   for (auto &op : module.getBody()->getOperations()) {
-    if (auto funcOp = mlir::dyn_cast<func::FuncOp>(op)) {
-      // Check if any gspmd operations exist.
-      mlir::WalkResult result = funcOp.getBody().walk([&](mlir::Operation *op) {
-        if (mlir::isa<mlir::stablehlo::CustomCallOp>(op)) {
-          auto customCall = mlir::dyn_cast<mlir::stablehlo::CustomCallOp>(op);
-          auto callTarget = customCall.getCallTargetName();
-
-          if (callTarget == "Sharding" ||
-              callTarget == "SPMDFullToShardShape" ||
-              callTarget == "SPMDShardToFullShape") {
-            return WalkResult::interrupt();
-          }
+    if (!mlir::isa<func::FuncOp>(op)) {
+      continue;
+    }
+
+    // Check if any gspmd operations exist.
+    func::FuncOp funcOp = mlir::cast<func::FuncOp>(op);
+    mlir::WalkResult result = funcOp.getBody().walk([&](mlir::Operation *op) {
+      if (mlir::isa<mlir::stablehlo::CustomCallOp>(op)) {
+        auto customCall = mlir::cast<mlir::stablehlo::CustomCallOp>(op);
+        auto callTarget = customCall.getCallTargetName();
+
+        if (callTarget == "Sharding" || callTarget == "SPMDFullToShardShape" ||
+            callTarget == "SPMDShardToFullShape") {
+          return WalkResult::interrupt();
         }
+      }
+
+      return WalkResult::advance();
+    });
 
-        return WalkResult::advance();
-      });
+    if (result.wasInterrupted()) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+// Check if the graph is solved.
+bool isGraphSolved(mlir::ModuleOp &module) {
+  for (auto &op : module.getBody()->getOperations()) {
+    if (!mlir::isa<func::FuncOp>(op)) {
+      continue;
+    }
 
-      if (result.wasInterrupted()) {
-        return true;
+    func::FuncOp funcOp = mlir::cast<func::FuncOp>(op);
+    mlir::WalkResult result = funcOp.getBody().walk([&](mlir::Operation *op) {
+      if (mlir::isa<mlir::sdy::ManualComputationOp>(op)) {
+        // If we find a manual computation op and it has its manual axes set, we
+        // consider the graph solved. (todo: tapspatel) - we need to generalize
+        // this to check for all manual computation ops, as we might have mixed
+        // graphs.
```

Comment:
`TODO` is usually written as uppercase and start it at newline. 
```suggestion
        // If we find a manual computation op and it has its manual axes set, we
        // consider the graph solved.
        // TODO (tapspatel) - we need to generalize this to check for all manual
        // computation ops, as we might have mixed graphs.
```